### PR TITLE
Implement TreeObject::load

### PR DESCRIPTION
**Pull request by AaronVanGeffen**

### CODE REVIEW

File: src/OpenLoco/src/Objects/TreeObject.cpp [hunk 1]
```diff
@@ -109,11 +112,101 @@ namespace OpenLoco
...
+        }
+
+        auto numVariantImages = numRotations * growth;
+        auto nextImageOffset = imageOffset;
+        for (auto variant = 0; variant < 6; variant++)
+        {
+            if ((var_3C & (1 << variant)) == 0)
+                continue;
+
+            sprites[0][variant] = imageOffset;
```
Comment (duncanspumpkin) [inline src/OpenLoco/src/Objects/TreeObject.cpp:142]: nextImageOffset.

File: src/OpenLoco/src/Objects/TreeObject.cpp [hunk 2]
```diff
@@ -109,11 +112,101 @@ namespace OpenLoco
...
+        {
+            for (auto variant = 0; variant < 6; variant++)
+            {
+                sprites[1][variant] = sprites[0][variant] + numPrimaryImages;
+            }
+
+            shadowImageOffset = numPrimaryImages * 2;
+        }
+        else
+            shadowImageOffset = numPrimaryImages;
```
Comment (duncanspumpkin) [inline src/OpenLoco/src/Objects/TreeObject.cpp:201]: your missing an if for L198 and this line. Presumably if hasShadow.

File: src/OpenLoco/src/Objects/TreeObject.cpp [hunk 3]
```diff
@@ -195,10 +195,15 @@ namespace OpenLoco
                 sprites[1][variant] = sprites[0][variant] + numPrimaryImages;
             }
 
-            shadowImageOffset = numPrimaryImages * 2;
+            nextImageOffset = numPrimaryImages * 2;
         }
         else
-            shadowImageOffset = numPrimaryImages;
+            nextImageOffset = numPrimaryImages;
```
Comment (duncanspumpkin) [inline src/OpenLoco/src/Objects/TreeObject.cpp:201]: can you put this behind {}. It just looks odd having one branch {}.

File: src/OpenLoco/src/Objects/TreeObject.cpp [hunk 4]
```diff
@@ -195,10 +195,15 @@ namespace OpenLoco
                 sprites[1][variant] = sprites[0][variant] + numPrimaryImages;
```
Comment (duncanspumpkin) [inline src/OpenLoco/src/Objects/TreeObject.cpp:195]: Should we maybe split the sprites array into snow and normal?

Comment (AaronVanGeffen) [reply src/OpenLoco/src/Objects/TreeObject.cpp:195]: I agree that would be more readable, yeah. Suggestions for new names? `normalSprites` and `snowSprites`?

File: src/OpenLoco/src/Objects/TreeObject.cpp (hunk 3 above)
Comment (AaronVanGeffen) [reply src/OpenLoco/src/Objects/TreeObject.cpp:201]:
```suggestion
        {
            nextImageOffset = numPrimaryImages;
        }
```

File: src/OpenLoco/src/Objects/TreeObject.cpp (hunk 4 above)
Comment (duncanspumpkin) [reply src/OpenLoco/src/Objects/TreeObject.cpp:195]: could just go `sprites` and `snowSprites`. Have a little check to see how over users of it use it.

Comment (AaronVanGeffen) [reply src/OpenLoco/src/Objects/TreeObject.cpp:195]: I imagine it'll only be mostly used by the paint routines, which aren't implemented yet.

**Review by AaronVanGeffen — COMMENTED** [review]
A question I had while implementing this, is whether the unload function as it is suffices? It's not setting the sprite arrays to 0, for starters.

**Comment by duncanspumpkin** [comment]
> A question I had while implementing this, is whether the unload function as it is suffices? It's not setting the sprite arrays to 0, for starters.

Should be fixed then. If the load function sets it the unload must clear it.

**Comment by AaronVanGeffen** [comment]
> Should be fixed then. If the load function sets it the unload must clear it.

My mistake, they were already being cleared. This much became clear after splitting the sprites array.

File: src/OpenLoco/src/Objects/TreeObject.cpp [hunk 5]
```diff
@@ -109,23 +112,97 @@ namespace OpenLoco
...
     {
         name = 0;
-        for (auto& spriteSeason : sprites)
+        for (auto& sprite : sprites)
         {
-            for (auto& sprite : spriteSeason)
-            {
-                sprite = 0;
-            }
+            sprite = 0;
```
Comment (duncanspumpkin) [inline src/OpenLoco/src/Objects/TreeObject.cpp:201]: sorry but could you use a std::fill.

File: src/OpenLoco/src/Objects/TreeObject.cpp [hunk 6]
```diff
@@ -109,23 +112,97 @@ namespace OpenLoco
...
+        }
+        else
+        {
+            nextImageOffset = numPrimaryImages;
+        }
+
+        if ((flags & TreeObjectFlags::hasShadow) != TreeObjectFlags::none)
+        {
+            shadowImageOffset = nextImageOffset;
+        }
```
Comment (duncanspumpkin) [inline src/OpenLoco/src/Objects/TreeObject.cpp:192]: Could you look at adding an assert like i have for vehicle on the image offsets. Would be nice to just be sure we allocated the correct amount and have used the correct amount.

**Comment by duncanspumpkin** [comment]
I've made a tweak so that it calculates the images a bit more like the vehicle object and fixed the assert for shadows.

**Comment by AaronVanGeffen** [comment]
Rebased on top of master. All looks good to me at this point, but pending review for merge.

**Review by duncanspumpkin — APPROVED** [review]

**Pull request merged**